src/python/grpcio/grpc/_cython/_cygrpc/records.pyx.pxi

```diff
@@ -462,6 +454,7 @@ cdef class Metadata:
       grpc_init()
       grpc_metadata_array_init(&self.c_metadata_array)
     metadata = list(metadata_iterable)
+    self._init_md_count = len(metadata)
```

Comment:
Why introduce this instance-scope field? Is it not redundant with `self.c_metadata_array.count`?

Reply:
No, it isn't because sometimes (server role) is Metadata object initialized with empty `metadata_iterable` parameter. In this case, the `self.c_metadata_array` (internal C structure) is maintained later by C code itself, including allocation of memory and unreferencing allocated memory (remember that python `Metadata` class is only proxy to C structure data). But sometimes (client role) are metadata passed directly through `metadata_iterable` parameter and in this case we have to unref allocated items manually. If we try to unref items according `self.c_metadata_array.count` sigsegv is invoked (server role), because C code is trying to unref unreferenced data. It's ok this way?

Reply:
Thank you for answering.

I tried to make sense of how your solution works, but I think several other problems in the code are getting in the way. If your change isn't urgent for you (it sounds like you're running fine with your change locally patched into your codebase) I'm going to take another day or so to look at possibly making a different change.

Reply:
Yeah, I think I'd like to make a different change - please let me know your thoughts about [pull request 12466](https://github.com/grpc/grpc/pull/12466)?